### DISCUSSION TITLE
axios get /app/installation works, but postapp/installations/<ID>/access_tokens NOT working

### DISCUSSION THREAD

**Discussion by flashflexpro** [community/community#66438] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

```
    async function refreshJWT() {
        {
            const now = Math.floor(Date.now() / 1000);
            jwtToken = jwt.sign({
                iat: now,
                exp: now + 60 * 10, // Token expires in 10 minutes
                iss: get_odmd_ghApp_ID(),
            }, secretKeyStr, {algorithm: 'RS256', encoding: 'utf8'});
        }
    }
    
    //works well
    const tmp = await axios.get(`${GITHUB_API_BASE_URL}/app/installations`, {
        headers: {
            Authorization: `Bearer ${jwtToken}`,
            Accept: 'application/vnd.github.v3+json',
            'X-GitHub-Api-Version': '2022-11-28'
        },
    }) as {    status: string, data:    [{id: string, access_tokens_url: string}]};
    
    for (let i = 0; i < tmp.data.length; i++) {
        await refreshJWT()
    
        const tmpUrl = tmp.data[i].access_tokens_url;
    
        try {
            /*
               always error:
                "message": "A JSON web token could not be decoded",
                "documentation_url": "https://docs.github.com/rest"
           */
            const tmp = await axios.post(tmpUrl, {
                headers: {
                    Authorization: `Bearer ${jwtToken}`,
                    Accept: 'application/vnd.github.v3+json',
                    'X-GitHub-Api-Version': '2022-11-28'
                },
            });
            console.log(tmp)
        } catch (e) {
            console.warn(e)
        }
    }

```

**Comment by flashflexpro** [comment]
AWS api/sdk is a lot better.

GH api is driving me crazy.

**Comment by divyam751** [comment, marked as answer]
The issue you're encountering with the POST request to the `access_tokens_url` might be related to the way you're setting the request headers and payload. The error message you provided, "A JSON web token could not be decoded," suggests that there might be a problem with the JWT token or how it's being used in the request.

Here are a couple of things to check and update in your code:

1. Set the correct content type: When making a POST request, you should specify the `Content-Type` header as 'application/json' to indicate that you're sending JSON data in the request body. It seems you've missed setting the content type.

2. Verify the JWT token: Ensure that the JWT token you're generating with `jwt.sign` is correctly signed and contains the required information. You may want to console log the `jwtToken` before making the POST request to verify its contents.

3. Use the correct syntax for axios: You should pass headers and data as separate objects when making the POST request. In your code, you've included the headers inside the data object, which is incorrect.

Here's an updated version of your code with these changes:

```javascript
async function refreshJWT() {
    const now = Math.floor(Date.now() / 1000);
    jwtToken = jwt.sign({
        iat: now,
        exp: now + 60 * 10, // Token expires in 10 minutes
        iss: get_odmd_ghApp_ID(),
    }, secretKeyStr, { algorithm: 'RS256', encoding: 'utf8' });
}

// Fetch installations
const installationsResponse = await axios.get(`${GITHUB_API_BASE_URL}/app/installations`, {
    headers: {
        Authorization: `Bearer ${jwtToken}`,
        Accept: 'application/vnd.github.v3+json',
        'X-GitHub-Api-Version': '2022-11-28'
    },
});

for (let i = 0; i < installationsResponse.data.length; i++) {
    await refreshJWT();
    const accessTokensUrl = installationsResponse.data[i].access_tokens_url;

    try {
        // Make POST request for access token
        const response = await axios.post(accessTokensUrl, {}, {
            headers: {
                Authorization: `Bearer ${jwtToken}`,
                Accept: 'application/vnd.github.v3+json',
                'X-GitHub-Api-Version': '2022-11-28',
                'Content-Type': 'application/json', // Set content type
            },
        });

        console.log(response.data); // Print the response data
    } catch (e) {
        console.warn(e);
    }
}
```

In this updated code:

- The `Content-Type` header is set to 'application/json' when making the POST request.
- The headers and data are passed as separate objects in the axios POST request.
- The response data is logged to the console for debugging purposes.

Make sure to check the JWT token generation and ensure it contains the correct information. If the issue persists, you may need to verify your JWT token signing process and the API endpoint you are trying to access.

**Reply by flashflexpro** [reply]
thanks for the quick help!
